wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -81,71 +96,151 @@ import           Pos.Util.Trace (Severity (Error))
+            -> Blund
+            -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta])
+mkPrefilter pw wId esk blund = blundToResolvedBlock (pw ^. walletNode) blund <&> \case
+    Nothing -> (M.empty, [])
+    Just rb -> prefilterBlock rb wId esk
+
+-- | Begin a restoration for a wallet that is already known. This is used
+-- to put an existing wallet back into a restoration state when something has
+-- gone wrong.
+restoreKnownWallet :: Kernel.PassiveWallet
+                   -> HD.HdRootId
+                   -> IO ()
+restoreKnownWallet pw rootId = do
+    let wId = WalletIdHdRnd rootId
+    lookupRestorationInfo pw wId >>= \case
+        -- Restart a pre-existing restoration
+        Just wri -> do
+            cancelRestoration  wri
+            restartRestoration wri
+
+        -- Start a new restoration of a seemingly up-to-date wallet.
+        Nothing -> Keystore.lookup wId (pw ^. walletKeystore) >>= \case
+            Nothing  -> return () -- TODO (@mn): raise an error
+            Just esk -> do
+                let prefilter = mkPrefilter pw wId esk
+                    wkey = (wId, keyToWalletDecrCredentials (KeyForRegular esk))
+
+                coreConfig <- getCoreConfig (pw ^. walletNode)
+                db <- getWalletSnapshot pw
+                case db ^. dbHdWallets . HD.hdWalletsRoots . at rootId of
+                    Nothing   -> return () -- TODO (@mn): this really shouldn't happen
+                    Just root ->
+                      let restart =
+                              withNodeState (pw ^. walletNode) (getWalletInitInfo coreConfig wkey) >>= \case
+                                  WalletCreate  _utxos     -> return ()
+                                  WalletRestore _utxos tgt ->
+                                    beginRestoration pw wId prefilter root tgt restart
+                      in restart
+
+beginRestoration  :: Kernel.PassiveWallet
+                  -> WalletId
+                  -> (Blund -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta]))
+                  -> HD.HdRoot
+                  -> BlockContext
+                  -> IO ()
+                  -> IO ()
+beginRestoration pw wId prefilter root tgt restart = do
+
+    let tgtTip  = tgt ^. bcHash   . fromDb
+        tgtSlot = tgt ^. bcSlotId . fromDb
+
+    -- Set the wallet's restoration information
+    slotCount <- getSlotCount (pw ^. walletNode)
+    progress <- newIORef $ WalletRestorationProgress
+                           { _wrpCurrentSlot = 0
+                           , _wrpTargetSlot  = flattenSlotId slotCount tgtSlot
+                           , _wrpThroughput  = MeasuredIn 0
+                           }
+    theTask <- newEmptyMVar
+
+    let restoreInfo = WalletRestorationInfo
+                      { _wriProgress = readIORef progress
+                      , _wriCancel   = readMVar theTask >>= cancel
+                      , _wriRestart  = restart
+                      }
+
+    addRestoration pw wId restoreInfo
```

Comment:
`restart` calls `begin` which calls `addRestoration`.. is that correct?

Reply:
Oh, I see. `addRestoration ` does more than just add it, it also cancels any existing. I guess that's okay; makes a tad worried about race conditions, if we could avoid replacing the entry altogether might be better. Probably minor though.